### DISCUSSION TITLE
Could not load file or assembly 'System.Runtime.CompilerServices.Unsafe, Version=4.0.4.1

### DISCUSSION THREAD

**Discussion by ilens** [jstedfast/MailKit#1822] [Q&A]
I have installed 4.8.0 via the NuGet. When I run within the Visual studio environment all works fine
When I make a installer with Inno Setup and include the dlls from the VS version, I get 
```
System.IO.FileLoadException: Could not load file or assembly 'System.Runtime.CompilerServices.Unsafe, Version=4.0.4.1, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040)
File name: 'System.Runtime.CompilerServices.Unsafe, Version=4.0.4.1, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'
   at System.Span`1..ctor(T[] array)
   at MimeKit.Utils.ValueStringBuilder..ctor(Int32 initialCapacity)
   at MimeKit.Utils.Rfc2047.EncodeAsBytes(FormatOptions options, Encoding charset, String text, Int32 startIndex, Int32 count, EncodeType type)
   at MimeKit.Utils.Rfc2047.EncodeText(FormatOptions options, Encoding charset, String text)
   at MimeKit.Header.EncodeUnstructuredHeader(ParserOptions options, FormatOptions format, Encoding encoding, String field, String value)
   at MimeKit.Header.EncodeAddressHeader(ParserOptions options, FormatOptions format, Encoding encoding, String field, String value)
   at MimeKit.Header.FormatRawValue(FormatOptions format, Encoding encoding, String value)
   at MimeKit.Header.SetValue(FormatOptions format, Encoding encoding, String value)
   at MimeKit.Header..ctor(Encoding encoding, HeaderId id, String value)
   at MimeKit.MimeMessage.CreateFromMailMessage(MailMessage message)
   at PCS_Emailer.clsEmailSender.send(String recipient, String subject, String text, List`1 attachments, Boolean msg, Boolean copyMe) in I:\Documents\Development\PCS\PCS_Emailer\Email\clsEmailSender.vb:line 229
   at PCS_Emailer.frmProgress.sendMessage(String msgType, String emailID) in I:\Documents\Development\PCS\PCS_Emailer\frmProgress.vb:line 181
   at PCS_Emailer.frmProgress.frmProgress_Shown(Object sender, EventArgs e) in I:\Documents\Development\PCS\PCS_Emailer\frmProgress.vb:line 63

WRN: Assembly binding logging is turned OFF.
To enable assembly bind failure logging, set the registry value [HKLM\Software\Microsoft\Fusion!EnableLog] (DWORD) to 1.
Note: There is some performance penalty associated with assembly bind failure logging.
To turn this feature off, remove the registry value [HKLM\Software\Microsoft\Fusion!EnableLog].
```

I notice the version reported in the VS reference is 4.0.40419 : - I have setSpecific Version to False but no luck.
![image](https://github.com/user-attachments/assets/ca4de917-97d9-44f0-8bb1-d287e0a2cf53)

Is the NuGet installer incorrect?

**Comment by jstedfast** [comment]
The System.Runtime.CompilerServices.Unsafe 6.0.0 nuget package contains an assembly with a different version than 6.0.0 (it's 4.something).

- What target framework are you using?
- What version of Visual Studio are you using?
- Is your build copying over the MimeKit.dll.config and MailKit.dll.config files?

You probably need to add an app.config file that contains the same content as the MimeKit.dll.config file that is bundled with the MimeKit nuget package.

**Comment by ilens** [comment]
Many thanks for the quick reply

 

I’m replacing system.net.mail with MailKit in an existing project.

 

Target framework 4.7.1

Visual studio 2022 community

 

My Inno Setup build looks like this…

 

Source: I:\Documents\Development\PCS\PCS_Emailer\bin\Debug\PCS_Emailer.exe; DestDir: {pf}\PCS_Utility; Flags: ignoreversion 

Source: I:\Documents\Development\PCS\PCS_Emailer\bin\Debug\PCS_Emailer.pdb; DestDir: {pf}\PCS_Utility; Flags: ignoreversion 

Source: I:\Documents\Development\PCS\PCS_Emailer\bin\Debug\Mailkit.dll; DestDir: {pf}\PCS_Utility; Flags: ignoreversion

Source: I:\Documents\Development\PCS\PCS_Emailer\bin\Debug\Mimekit.dll; DestDir: {pf}\PCS_Utility; Flags: ignoreversion

 

Source: I:\Documents\Development\PCS\PCS_Emailer\bin\Debug\Mailkit.dll.config; DestDir: {pf}\PCS_Utility; Flags: ignoreversion

Source: I:\Documents\Development\PCS\PCS_Emailer\bin\Debug\Mimekit.dll.config; DestDir: {pf}\PCS_Utility; Flags: ignoreversion

 

Source: I:\Documents\Development\PCS\PCS_Emailer\bin\Debug\System.Buffers.dll; DestDir: {pf}\PCS_Utility; Flags: ignoreversion

Source: I:\Documents\Development\PCS\PCS_Emailer\bin\Debug\System.Formats.Asn1.dll; DestDir: {pf}\PCS_Utility; Flags: ignoreversion

Source: I:\Documents\Development\PCS\PCS_Emailer\bin\Debug\system.Memory.dll; DestDir: {pf}\PCS_Utility; Flags: ignoreversion

Source: I:\Documents\Development\PCS\PCS_Emailer\bin\Debug\System.Numerics.Vectors.dll; DestDir: {pf}\PCS_Utility; Flags: ignoreversion

Source: I:\Documents\Development\PCS\PCS_Emailer\bin\Debug\System.Runtime.CompilerServices.Unsafe.dll; DestDir: {pf}\PCS_Utility; Flags: ignoreversion

Source: I:\Documents\Development\PCS\PCS_Emailer\bin\Debug\System.Threading.Tasks.Extensions.dll; DestDir: {pf}\PCS_Utility; Flags: ignoreversion

Source: I:\Documents\Development\PCS\PCS_Emailer\bin\Debug\System.ValueTuple.dll; DestDir: {pf}\PCS_Utility; Flags: ignoreversion

**Reply by jstedfast** [reply]
Okay, so this is a known issue with .NET Framework. Not sure which version of Visual Studio 2022 you are using (thought the issue had been fixed, but maybe not?).

You need to create an app.config file to solve this issue. See the FAQ for details: https://github.com/jstedfast/MailKit/blob/master/FAQ.md#type-load-exception

**Comment by ilens** [comment]
The app config already has that <runtime> code.

**Reply by jstedfast** [reply]
Is your app.config getting copied into your app bundle?

The app.config file is the only thing you need to solve this issue. If that still doesn't work, then you'll probably need to go and ask the dotnet runtime team for help.

**Comment by ilens** [comment]
Ah, I’m outside my competence envelope here - where is the app bundle and how can I tell if it’s in it?

 

(The <runtime> code is in the Mailkit.dll.config as well, and I copied that in my installer)

 

Is there anywhere I can get System.Runtime.CompilerServices.Unsafe, Version=4.0.4.1, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a ?

**Reply by jstedfast** [reply]
When I say your "app bundle", I'm just asking if the app.config file gets installed by your installer. It didn't look like it to me, that's why I asked.

The nuget versions of System.Runtime.CompilerServices.Unsafe do not match their assembly versions (hence the problem). If you go to https://www.nuget.org/packages/System.Runtime.CompilerServices.Unsafe/6.0.0 - the 6.0.0 nuget package contains the 4.0.4.1 assembly version.

**Comment by ilens** [comment]
Thanks for your help – I think I’m going to have to give up. It’s a shame because MailKit looks a really good library
